### DISCUSSION TITLE
GO Source class and file naminng and directory layout

### DISCUSSION THREAD

**Discussion by oleg68** [GrandOrgue/grandorgue#793] [Developers]
Now there are two source directories with hundreds of source files: `core` and `grandorgue`. It is very hard to navigate across such large number of unclassified files.

The source files have three different prefixes: `GO`, `GOrgue` and `GrandOrgue`. Some files do not have a prefix. I couldn't catch a common idea, when each preffix is used. Because I never remember the actual prefix of a certain file, it is hard to find a particular file.

The contents of `core` is built into the `libGrandOrgueCore`  dynamic library. This library is used also by `GrandOrgueTool`. But seems GrandOrgueTool uses less than 5% of it's contents. So all other 95% are candidates to move to grandorgue.

My suggestions are

1. To select one preffix among three ones.
2. To rename all files to use this prefix
3. To create some subdirectories under `core` and `grandorgue` and to distribute source files across them according to their domain. For example, `audio`, `audio/scd`, `gui`, `settings` and so on.
    This might be a long process. I have already started making this in #765 for `core/threading` subrirectory. May we continue?

**Comment by larspalo** [comment]
golib is also used by for instance perftest, that incidentally seems to be non functional because of your restart addition:

[100%] Linking CXX executable perftest
/usr/bin/ld: libgolib.a(GrandOrgueFrame.cpp.o): in function `GOrgueFrame::OnSettings(wxCommandEvent&)':
/home/lars/Sources/github/grandorgue/src/grandorgue/GrandOrgueFrame.cpp:932: undefined reference to `GOrgueApp::SetRestart()'
collect2: error: ld returned 1 exit status

Also note the irony in naming of the file GrandOrgueFrame compared to the class name GOrgueFrame.

**Comment by midimusic** [comment]
I have tried on several occasions to understand the GO source files and have never got very far. Mainly because there are no comments in the files and thus their intention is obscure. At this time I have not even discovered which file is the "base" on which all the others depend.
Could it be the GrandOrgueFrame.cpp mentioned above?

I am also trying to understand and use wxWidgets in a separate project. 

It would be a very good idea to go through all of the source files and at least add a comment stating its purpose at the beginning of each. The "nonsense" comment at the start of every file should be deleted and if necessary put into one separate "copyright" file.

I thus fully agree that there is a lot of room for improving the logic and understandability of the GO source files. This will make it much easier to make sorely needed future improvements.

csw900

**Comment by oleg68** [comment]
@larspalo 

> /home/lars/Sources/github/grandorgue/src/grandorgue/GrandOrgueFrame.cpp:932: undefined reference to GOrgueApp::SetRestart()'

Which source code of GOrgueApp are you using? The [standard one](https://github.com/GrandOrgue/grandorgue/blob/master/src/grandorgue/GrandOrgue.h) declares SetRestart() at the line 59.

The file name GrandOrgue.h also differs from the class name GOrgueApp. It is a hell!

Let choose the single prefix for the all classes and files and I start renaming them.

**Reply by larspalo** [reply]
The GO source has been worked on by so many people that no single approach has been fully realized, thus many things are indeed "messy". Also, some work has been done by people not seeing the greater picture...

The last serious effort was mainly to have a GUI separation from the backend functionality (a kind of MVC approach). That's why there's the GOGUI and GOrgue file/class separation. I suspect the GO prefix was chosen for specifying some special functions like Sound because of a shorter file naming (GOSound). Then the older files that wasn't yet reworked could had any kind of naming really, and the same went for the helping classes that fixed certain issues.

The naming scheme and possible folder organization can really be anything we agree upon, as long as we have a clear and systematic idea behind it, it won't be worse than what it is today... That's of course only my opinion.

I have to look into the reason why perftest linking fail. I'm not even sure why the GrandOrgueFrame is pulled in. I'll have to track what changes have been done since rev 2341 if you don't know already what's causing the issue.

**Comment by oleg68** [comment]
@larspalo I added GOrgueApp::SetRestart() in https://github.com/GrandOrgue/grandorgue/commit/588c348c98faee7f6f9532bf82b0b45e67137b2b

But I don't know why is it causing any compiling issues.

**Reply by larspalo** [reply]
The compiling of perftest is fine, it's in the linking to golib that it fails with the message above.

**Reply by oleg68** [reply]
May be you are trying to take an old lib version?

**Comment by oleg68** [comment]
@larspalo separating GUI from the backend is a super thing! I'd use the `grandorgue/gui` subdirectory for it.

**Reply by oleg68** [reply]
I suggest to use ’GO' prefix for all classes. 'GOGUI' may continue existing because it is two prefixes: 'GO' + 'GUI', but 'GOrgue' and 'GrandOrgue' should be renamed to 'GO'. The files containing the classes should be renamed according their classnames.

@larspalo do you agree with 'GO'?

**Reply by larspalo** [reply]
Yes, I think that the GO prefix is a good start. For the GUI files you still want a subdirectory too? And a subdirectory for all sound related files? And keeping the core and grandorgue separation of source files for the lib?

This requires quite a lot of changes to realize... Perhaps it's best to do this re-organization after the release?

**Reply by oleg68** [reply]
> Yes, I think that the GO prefix is a good start. 
> 
> This requires quite a lot of changes to realize... Perhaps it's best to do this re-organization after the release?

Yes, it is. 3.4.2. I've created #816 for renaming

> For the GUI files you still want a subdirectory too? And a subdirectory for all sound related files? 

Yes, it is a good idea. 

Lets discuss here the target directory structure and  then create an issue. But it should be done in a separate PR otherwise it is hard to review.

> And keeping the core and grandorgue separation of source files for the lib?

I think, yes. May be we'll return to this question later.

**Comment by oleg68** [comment]
I suggest to reorganize GO source code directory structure piece by piece. So most files rest at their places, but move only files that are evident.

First of all, I'd like to create the following subdirectories

- src
    - core
        - config
        - settings
    - grandorgue
        - gui
        - midi
        - settings
        - sound
            - scd
            - ports

**Reply by larspalo** [reply]
What is /sound/scd intended for? SCD has no obvious meaning in this context, as far as I can translate it... Google it just for fun!

Eventually, I think a separation of the gui elements of the organ and the gui of GO in itself would be a good idea.

**Reply by oleg68** [reply]
@larspalo 

> What is /sound/scd intended for? 

Scheduler. May be to name it `scheduler` instead of `scd`. I'd like to put there GOSOundThread, GOSoundScheduler, GOSoundWorkItem and all derived classes (that are heavy using mutices).

My another idea is to have the same directory layout under `core`.

**Reply by larspalo** [reply]
> May be to name it `scheduler` instead of `scd`.

Yes, that's much more readable.

**Comment by oleg68** [comment]
> golib is also used by for instance perftest, that incidentally seems to be non functional because of your restart addition:

I think we have to include building perftest in the same cmake build as like as GrandOrgue and GrandOrgueTool for making sure that it can be compiled.

**Reply by larspalo** [reply]
That could be a very good idea, even though it's mainly targeted towards developers to evaluate how some changes will affect the available polyphony.

It might be possible that, in the future, we'll also ship other applications (that possibly depend on GO or it's lib) together with GO that could be useful for sampleset creators or those wanting to modify samplesets. Then this would be an excellent approach of building them all together just like the GrandOrgueTool.

**Comment by oleg68** [comment]
Also it would be good to move GrandOrgueTool and perftest outside the grandorgue directory.

**Reply by larspalo** [reply]
Yes, maybe to a `tools` directory or something similar, since they are separate from the GO app but uses the same core lib.

**Comment by oleg68** [comment]
Yes, it may. GrandOrgueTool and peftest have a little of souce code so they may coexist in the same directory `tools`.

**Comment by oleg68** [comment]
So, for now:

- src
    - core
        - config
        - gui
        - settings
        - sound
            - scheduler
    - grandorgue
        - gui
        - midi
        - settings
        - sound
            - scheduler
            - ports
    - tools

Some notes:
1. I put `core/sound/scheduler` for only three files: `GOSoudTouchWorkItem.cpp`, `GOSoudTouchWorkItem.h` and `GOSoudWorkItem.h`. I don't know now, may we move them to `grandorgue/sound/scheduler` or not.
2. I'd like to make a MVC approach inside the `gui`: `gui/model` and 'gui/view`. But now GOGU* classes are close to the model, but vave some 'controller' or 'view' methods. So for now I suggest to move all gui-related classes together and to divide them into subdirectories in the future.
3. Where should we put the `GOApp` and `GoFrame` (may be rename it to `GOMainPanel` or `GOControlPanel`) classes? I suggest to `grandorgue/gui`.

@larspalo If you agree with this structure, I'll create an issue for 3.4.2.

**Reply by larspalo** [reply]
1. If they have a purpose in the library I think they should be left within `core` otherwise they should be moved to within `grandorgue`, I think
2. We can group them together for now, I think
3. I'm not sure that we should put **everything** into the `gui`. Personally, I'd like a separation of all the files that relate to the GOApp itself from the GOGUI files that relate to organ specific rendition (some of which possibly in the future should be moved to core instead if they could be useful for other programs). But if it's too much right now we could leave all the basic app related files in the `grandorgue` and just move the GOGUI files to `grandorgue/gui`.

This is a start and we'll likely have to adjust many things in the future anyway.

**Comment by oleg68** [comment]
1. I cann't answer now. So now we will left them in `core`
2. OK
3. OK

> This is a start and we'll likely have to adjust many things in the future anyway.

Indeed.

So I'm creating an issue...

**Comment by oleg68** [comment]
#834

**Comment by oleg68** [comment]
Working on MIDI setting, I've noted, that 
1. there are two different kinds of classes in the `src/grandorgue/settings`
    1. The config objects that keep the GrandOrgue settings. The main one is `GOSettings`
    2. SettingsDialog and the UI elements are used by SettingsDialog
2. Most of GrandOrgue source code depends on the config objects but not on the settings UI.
3. Most of settings UI classes do not have the `GO` preffix.

I'd like to do some reorganising:
1. To split `settings` to two different directories: `config` and `settings-ui`
2. To rename `GOSettings` to `GOConfig` and move to `config`
3. To add `GO` prefix to all `Settings` classes.

@larspalo do you agree with this reorganising? If so, I'll make a PR.

**Comment by oleg68** [comment]
Before working with help buttons on Midi Event dialogs #1001 and on the Program Settings #1003 I'd like to make some directory structure changes:

- `src/grandorgue/dialogs`
    - `common`: there will be a base class for all GO dialogs with help and page navigation code
    - `midi-event`: all files from `src/grandorgue/midi/*Dialog*`.
    - `organ`: GOOrganDialog
    - `settings`: all files from `src/grandorgue/settings-gui`.

Next, if a dialog has pages, then only the main class should have `Dialog` in it's name. All other classes  (MIDIEventKeyDialog, MIDIEventRecvDialog, MIDIEventSendDialog) should be renamed.

Of cause, all MIDIEvent* classes should receive the `GO` prefix.

@larspalo Do you agree with these changes?
